src/extensions/Statiq.ApiClient/ReadApiWithClient.cs

```diff
@@ -0,0 +1,149 @@
+        public ReadApiWithClient<TClient> WithClientInitialization(Action<IDocument, IExecutionContext, TClient> init)
+        {
+            if (init == null)
+            {
+                throw new ArgumentException("Argument is null or empty", nameof(init));
+            }
+
+            _init = init.ThrowIfNull(nameof(init));
+            return this;
+        }
+
+        /// <summary>
+        /// Submits the API client name (for logging).
+        /// </summary>
+        /// <param name="name">Client name.</param>
+        /// <returns>The current module instance.</returns>
+        public ReadApiWithClient<TClient> WithClientName(string name)
+        {
+            _clientName = name.ThrowIfNullOrWhiteSpace(nameof(name));
+            return this;
+        }
+
+        /// <summary>
+        /// Submits a request to the API client. This allows you to incorporate data from the execution context in your request.
+        /// </summary>
+        /// <param name="key">The metadata key in which to store the return value of the request function.</param>
+        /// <param name="request">A function with the request to make.</param>
+        /// <returns>The current module instance.</returns>
+        public ReadApiWithClient<TClient> WithRequest(string key, Func<IExecutionContext, TClient, object> request)
+        {
+            key.ThrowIfNullOrWhiteSpace(nameof(key));
+            request.ThrowIfNull(nameof(request));
+
+            _requests[key] = (doc, ctx, client) => request(ctx, client);
+            return this;
+        }
+
+        /// <summary>
+        /// Submits a request to the API. This allows you to incorporate data from the execution context and current document in your request.
+        /// </summary>
+        /// <param name="key">The metadata key in which to store the return value of the request function.</param>
+        /// <param name="request">A function with the request to make.</param>
+        /// <returns>The current module instance.</returns>
+        public ReadApiWithClient<TClient> WithRequest(string key, Func<IDocument, IExecutionContext, TClient, object> request)
+        {
+            key.ThrowIfNullOrWhiteSpace(nameof(key));
+
+            _requests[key] = request.ThrowIfNull(nameof(request));
+            return this;
+        }
+
+        /// <inheritdoc/>
+        protected override IEnumerable<IDocument> ExecuteInput(IDocument input, IExecutionContext context)
+        {
+            _init?.Invoke(input, context, _client);
+            ConcurrentDictionary<string, object> results = new ConcurrentDictionary<string, object>();
+            System.Threading.Tasks.Parallel.ForEach(_requests, request =>
+            {
+                context.LogDebug("Submitting {0} {1} request for {2}", request.Key, _clientName, input.ToSafeDisplayString());
+                try
+                {
+                    results[request.Key] = request.Value(input, context, _client);
```

Comment:
We should probably check for null (using `value is null` instead of `value == null` - I switched all the null checks to that convention recently). If a null result, don't add to the dictionary. Then check at the end whether the results dictionary actually has any results and avoid the clone and just return the original document if so.

Reply:
It will be done.